### ISSUE TITLE
Feature Implement

### ISSUE THREAD

**Issue by AstroProjection** [Obsession-Boutique/Joannie-Store-Shop#1]
Fixing the issue of Delete Product Review as discussed on https://github.com/bradtraversy/proshop_mern/issues/141

Please update your 'master' branch to your latest changes and push it to origin

**Comment by talmax1124** [comment]
I don't see the PR

**Comment by AstroProjection** [comment]
Will do it, but I wanted to latest version of your main branch so that I can work off of that

**Comment by talmax1124** [comment]
Ok, just did

**Comment by AstroProjection** [comment]
I don't have access to push to your repository

**Comment by talmax1124** [comment]
Let me do that then

**Comment by talmax1124** [comment]
Try now

**Comment by AstroProjection** [comment]
https://github.com/Obsession-Boutique/Joannie-Store-Shop/pull/2

Verify and let me know.

**Comment by talmax1124** [comment]
It does nothing: [0] DELETE /api/products/6089626dcc7322a7134fa879/reviews 201 323.964 ms - 28

**Comment by talmax1124** [comment]
<img width="1182" alt="Screen Shot 2021-04-28 at 10 21 37 AM" src="https://user-images.githubusercontent.com/61528430/116420153-ba61a480-a80b-11eb-91c4-49a114097022.png">

**Comment by AstroProjection** [comment]
Could you send a pic of the payload that's being sent on the delete request as well as the headers which are being sent

**Comment by talmax1124** [comment]
<img width="754" alt="Screen Shot 2021-04-28 at 10 33 21 AM" src="https://user-images.githubusercontent.com/61528430/116421812-2db7e600-a80d-11eb-8457-63713b86e475.png">
<img width="757" alt="Screen Shot 2021-04-28 at 10 33 40 AM" src="https://user-images.githubusercontent.com/61528430/116421836-33adc700-a80d-11eb-85d8-3d91f06e4254.png">
<img width="756" alt="Screen Shot 2021-04-28 at 10 33 59 AM" src="https://user-images.githubusercontent.com/61528430/116421890-40321f80-a80d-11eb-96f8-ac1909876807.png">
![Uploading Screen Shot 2021-04-28 at 10.34.12 AM.png…]()

**Comment by AstroProjection** [comment]
Could you you send an screenshot of the payload, it should be the review data that we're sending from the frontend to the back

**Comment by talmax1124** [comment]
Where do I find the payload?

**Comment by talmax1124** [comment]
https://user-images.githubusercontent.com/61528430/116430124-89d23880-a814-11eb-9785-cd2bec2e59a0.mov

**Comment by talmax1124** [comment]
Does this help?

**Comment by talmax1124** [comment]
@AstroProjection Once we finish this, the next thing I want to integrate is Options. I created a issue about this in the proshop repo as well

**Comment by AstroProjection** [comment]
https://github.com/Obsession-Boutique/Joannie-Store-Shop/pull/3 - this PR should resolve the issues

**Comment by talmax1124** [comment]
Yes it is working now

**Comment by talmax1124** [comment]
Ok, I implemented coupons. I think this may be a stupid fix, but when I do multiplication by 0.3 which is supposed to be 30% gives me a bunch of numbers after it. You know for example 3.2993448 when I have .toFixed(2). Any fixes you can think of?

**Comment by talmax1124** [comment]
You want me to explain where?

**Comment by AstroProjection** [comment]
I'm not sure I understand the question, the float.toFixed(2) isn't working?

**Comment by talmax1124** [comment]
Exactly. Ok, so I am using the files cartAction: 
export const addCoupon = (coupon) => (dispatch) => {
  let couponDiscount = 0;

  // TODO: Get couponDiscount value from db

  if (coupon.toLowerCase() === "familydiscount") {
    couponDiscount = 3;
  }

  if (coupon.toLowerCase() === "militarydiscount") {
    couponDiscount = 2;
  }

  if (coupon.toLowerCase() === "sacoleiraspecialty") {
    couponDiscount = 4;
  }

  dispatch({
    type: CART_ADD_COUPON,
    payload: { coupon, couponDiscount },
  });

  localStorage.setItem("coupon", JSON.stringify(coupon));
  localStorage.setItem("couponDiscount", JSON.stringify(couponDiscount));
};

I added a coupon constant in the cart constant: export const CART_ADD_COUPON = "CART_ADD_COUPON";

Added to reducer: 
 case CART_ADD_COUPON:
      return {
        ...state,
        coupon: action.payload.coupon,
        couponDiscount: action.payload.couponDiscount,
      };

and in placeorderscreen is: 

if (cart.couponDiscount) {
    cart.totalPrice *= Number(cart.couponDiscount).toFixed(2);
  }

**Comment by talmax1124** [comment]
The major component is in placeorderscreen. So I have the product price for example $7. Then the fee which is like $3.30. Then I get the value from the action and multiply by the cart total. When I add .toFixed(2) or even without toFixes(2) it does nothing. Doesn't constrain to two variables

**Comment by AstroProjection** [comment]
Add toFixed to the final cart.totalPrice

**Comment by talmax1124** [comment]
I did this:   if (cart.couponDiscount) {
    cart.totalPrice.toFixed(2) *= Number(cart.couponDiscount).toFixed(2);
  } 
and in terminal got: SyntaxError: /Users/carlitos/shop/frontend/src/screens/PlaceOrderScreen.js: Invalid left-hand side in assignment expression (75:4)

**Comment by talmax1124** [comment]
I don't understand why this is doing it because everything is using .toFixed(2) in terms of the totalPrice and all that

**Comment by AstroProjection** [comment]
```
if (cart.couponDiscount) {
cart.totalPrice *= Number(cart.couponDiscount).toFixed(2);
cart.totalPrice = cart.totalPrice.toFixed(2);
}
```

You are multiplying the previous totalPrice with a new float value which will make it behave like that again

**Comment by talmax1124** [comment]
Yay, it is working! Thanks. You have been so helpful

**Comment by talmax1124** [comment]
I hate to keep bothering you but can we work on a vscode live share session. I started this feature but I was having multiple inputs and it was not passing to cartScreen. It's options

**Comment by talmax1124** [comment]
I really appreciate everything you have done for me. I have been trying to get help from other people and all I gotten is I can't help you or I'm busy. So I appreciate you for helping me. It means alot. I know words may not mean a-lot, I would pay you if I had money. Honestly. As a 16 year old, I got into programming last year, and it has been a journey. A great one.

**Comment by talmax1124** [comment]
I also found a bug in the reviews. It deleted the review, but when I'm seeing the word, it says 1 review when there is no review

**Comment by AstroProjection** [comment]
Yeah sure, I can help you out, but i'm a bit busy during the day, it's 10AM where I live but in the evening I can try and help you out

**Comment by talmax1124** [comment]
Ok, thanks! You really are a lifesaver. It's 12:17 in the night for me. But at the time you messaged me is fine

**Comment by talmax1124** [comment]
Can you reply back here when you are available so I can come to this page. I get email notifications when someone replies. Thanks

**Comment by talmax1124** [comment]
I think I am doing the calculations wrong, for the coupon I mean. I am getting the 30% of the total and not the total minus the 30%. Any fixes?

**Comment by talmax1124** [comment]
Ok, I did this:   if (cart.couponDiscount) {
    var tempDisc = Number(cart.couponDiscount) * cart.totalPrice
    cart.totalPrice -= tempDisc
    cart.totalPrice = cart.totalPrice.toFixed(2);
  }

You think this is efficient?

**Comment by talmax1124** [comment]
Are you available now?

**Comment by talmax1124** [comment]
What time is it for you now?

**Comment by AstroProjection** [comment]
> I think I am doing the calculations wrong, for the coupon I mean. I am getting the 30% of the total and not the total minus the 30%. Any fixes?

Total * 0.7 will give you the total after 30% discount

> Ok, I did this: if (cart.couponDiscount) {
> var tempDisc = Number(cart.couponDiscount) * cart.totalPrice
> cart.totalPrice -= tempDisc
> cart.totalPrice = cart.totalPrice.toFixed(2);
> }
> 
> You think this is efficient?

It should be fine

**Comment by talmax1124** [comment]
I got that working!

**Comment by talmax1124** [comment]
https://prod.liveshare.vsengsaas.visualstudio.com/join?79EF46F7F6DD3CA7348F4499315E061DF29C Here is my live share link

**Comment by talmax1124** [comment]
Can we continue chatting in discussions?

**Comment by talmax1124** [comment]
> It's 8:30, it's been a busy day

It's been a busy morning for me too!

**Comment by talmax1124** [comment]
@AstroProjection Discussion Link: https://github.com/Obsession-Boutique/Joannie-Store-Shop/discussions/5

**Comment by talmax1124** [comment]
Will you be able to help me today or tomorrow?

**Comment by talmax1124** [comment]
Good Morning!

**Comment by talmax1124** [comment]
Good Morning to you! When will you be able to assist?

**Comment by AstroProjection** [comment]
Hey good morning, again busy during the day but ask your queries and if I can help i'll give you suggestions

**Comment by talmax1124** [comment]
When are you going to be available to help in the programming process? This
part I am really confused and have tried multiple times. The other things
were things I had integrated in the backend but needed help in the
frontend.

Whenever you can help me out like scope it out, I would appreciate it.

**Comment by talmax1124** [comment]
The thing that I need your help with is adding options and qty in products. Sort of like this: 

https://user-images.githubusercontent.com/61528430/116652632-98733980-a953-11eb-8e6f-148bdc797f37.mov

@AstroProjection

**Comment by AstroProjection** [comment]
Get started on it and I'll look into it later in the day, make a new branch from your master and make your changes on that

**Comment by talmax1124** [comment]
ok

**Comment by talmax1124** [comment]
I just pushed it to the branch

**Comment by talmax1124** [comment]
Are you going to be available now? I updated the branch again. Can we continue in the discussions tab? @AstroProjection

**Comment by talmax1124** [comment]
Is it 7pm for you? Can you see the new branch? I called it options

**Comment by AstroProjection** [comment]
Hey, yea I see the branch, what are you trying to do?

**Comment by talmax1124** [comment]
Do you see the video, I posted it in this comment: https://github.com/Obsession-Boutique/Joannie-Store-Shop/issues/1#issuecomment-829849242

**Comment by talmax1124** [comment]
Do you get what I mean?

**Comment by talmax1124** [comment]
Any suggestions?

**Comment by talmax1124** [comment]
Can we set a time to work on it?

**Comment by AstroProjection** [comment]
Look for some suggestions online for the same layout you want to implement and take the code from there and tweek it with your values and see how it goes, that layout should be available online

For the code you have already setup, just keep a state variable which tells you which size is selected and only display those values when the user has checked one of the checkboxes, using a Checkbox component from whatever frontend library you're using.

I'm busy working on some stuff but i'll try and help you out when I can

**Comment by talmax1124** [comment]
I have been at this over a week trying to implement just this, thats why I
have been persistent in seeing if someone can help me out.

**Comment by talmax1124** [comment]
& thanks for everything you have done for me. I know I'm asking for alot and I'm sorry

**Comment by AstroProjection** [comment]
You're welcome, I know it's a little frustrating when you're just learning and everything seems very foreign. But keep that branch of pickingoptions updated with your latest changes I'll look into it now and then and see if I can help you get this feature running

**Comment by talmax1124** [comment]
> You're welcome, I know it's a little frustrating when you're just learning and everything seems very foreign. But keep that branch of pickingoptions updated with your latest changes I'll look into it now and then and see if I can help you get this feature running

It is frustrating, especially when you have your mind set to something and I can't accomplish it. I just wish as time goes on, I increase my thinking process and problem solving so that I can do things like this. I'm sorry if I seem pushy or persistent. I really am. I know you have things to do too. So I'm sorry.

**Comment by talmax1124** [comment]
& I know I'm asking alot, but this is something that I have asked various times, and it's how to make a single image appear in the homescreen and a carousel or a multi image display like amazon in the product-screen. I keep getting responses of making it into an array, but I am not understanding on how to do that. I hate asking you this especially with everything you have done, but can you teach me how to do that or include it in the PR? @AstroProjection

**Comment by AstroProjection** [comment]
> > You're welcome, I know it's a little frustrating when you're just learning and everything seems very foreign. But keep that branch of pickingoptions updated with your latest changes I'll look into it now and then and see if I can help you get this feature running
> 
> It is frustrating, especially when you have your mind set to something and I can't accomplish it. I just wish as time goes on, I increase my thinking process and problem solving so that I can do things like this. I'm sorry if I seem pushy or persistent. I really am. I know you have things to do too. So I'm sorry.

Yeah, you need to make a habit of understanding the problem and searching for the right solution online, that's how you get better.



> & I know I'm asking alot, but this is something that I have asked various times, and it's how to make a single image appear in the homescreen and a carousel or a multi image display like amazon in the product-screen. I keep getting responses of making it into an array, but I am not understanding on how to do that. I hate asking you this especially with everything you have done, but can you teach me how to do that or include it in the PR? @AstroProjection
> 
> I mentioned it in this issue: [bradtraversy/proshop_mern#135](https://github.com/bradtraversy/proshop_mern/issues/135)
> as well as UDEMY course

by using an array what they mean is that you would create a obj like
obj = {
 imgURL:'url...'
}

or even an array ( i.e array1 = ['url1','url2',....] )

Then you would do a 
`
array1.map(url=>{
   return <img src={url} /> // or whatever component you're using to show the image
});

`

That's pretty much it

**Comment by talmax1124** [comment]
Do I need to create a component?

> Yeah, you need to make a habit of understanding the problem and searching for the right solution online, that's how you get better.

I have been looking online for like 5 hours trying to look and trying to implement. So you were like my last hope. I have been looking at the google, github repo's, basically anything. I found alot of cool repo's with features I like, but it just didn't align alongside with what I was doing, especially since they used firebase.

**Comment by talmax1124** [comment]
Has creating that feature been difficult for you? The options thing? @AstroProjection

**Comment by AstroProjection** [comment]
I don't think it will be difficult, but it also requires a good way to handle making sure you're getting the right count of the stock and stuff, to keep it simple you can just use the number it shows and then just verify if it worked from the backend call.

The frontend part is just about conditonally showing a different list, 

Like I'd said earlier, make a state variable and keep the user selected option (i.e , L,XL...etc )[singular]
Then just display the list according to the selected option

**Comment by talmax1124** [comment]
> I don't think it will be difficult, but it also requires a good way to handle making sure you're getting the right count of the stock and stuff, to keep it simple you can just use the numbers from the initial backend that's getting the product, then just verify if the purchase worked when you make the backend purchase request.
> 
> 
> 
> The frontend part is just about conditonally showing a different list, 
> 
> 
> 
> Like I'd said earlier, make a state variable and keep the user selected option (i.e , L,XL...etc )[singular]
> 
> Then just display the list according to the selected option
-----
> You're welcome, I know it's a little frustrating when you're just learning and everything seems very foreign. But keep that branch of pickingoptions updated with your latest changes I'll look into it now and then and see if I can help you get this feature running

I thought you were helping me with a PR. I'm sorry if I misunderstood. I really don't know how to accomplish it. It's fine though. Sorry, if I'm stating it wrong. My primary language is Spanish, so I'm trying to phrase everything good. @AstroProjection

**Comment by AstroProjection** [comment]
Yeah at the moment I'm just a little caught up to spend some time going through the code but the solution is fairly straight forward and I think you should be able to accomplish it.

When I was learning, hands out trial and error was one of the main reasons I am even so comfortable in javascript right now, and I think you'll benefit greatly from just trying out different methods.

Spend more time learning react if you're finding the front-end a little complicated, go through reactjs.org

And if that's too much, just skim through mdn https://developer.mozilla.org/en-US/docs/Web/JavaScript

**Comment by talmax1124** [comment]
It's fine, I'll just discard that feature, because I'm in school most of the time. I won't have enough time to implement it by tomorrow. @AstroProjection

**Comment by talmax1124** [comment]
I kinda shut down when I get stressed and can't accomplish something. So, it's the only reason I was asking for that assistance

**Comment by AstroProjection** [comment]
Yea I can imagine, don't rush it, take it slow and make incremental changes, and if you find you aren't understanding how something in javascript/html/css works, just go on a tangent and look into it. Don't rush to finish it quickly. You'll find just a couple hours a day just looking into a feature will improve your knowledge and fluency in a month or two

**Comment by talmax1124** [comment]
Thanks for the advice. I'll see if someone in Udemy can help me in this.

**Comment by talmax1124** [comment]
Hi! I still need help in this. Would you mind just helping me finish this one thing?

**Comment by talmax1124** [comment]
Hey @AstroProjection, Still couldn't accomplish the feature. Could you please help me out in the sizes and the images carousel for products?

**Comment by AstroProjection** [comment]
Yeah sure, how much have you managed to get done?

**Comment by talmax1124** [comment]
I had to delete it because it was errors after errors. I am restarting the code for this feature again, because I'm doing something wrong. I was working on the images first.

**Comment by talmax1124** [comment]
Hey @AstroProjection  ! Can you help me implement something real quick. I have had enough people creating fake profiles and purchasing with the fake profile. Will you be able to help me implement email verification alongside google bot verification.

**Comment by AstroProjection** [comment]
Sure, are you sending emails out from your platform?

**Comment by talmax1124** [comment]
> Sure, are you sending emails out from your platform?
> 
> 

What do you mean by that? I just want to be able to make the user verify there email via a email once they register. If they verified it then they can login or maybe make a purchase. I don't know yet. @AstroProjection 

I have been looking at services that does this but I don't know which one to choose and their documentation is horrible.

**Comment by AstroProjection** [comment]
I meant are you sending emails when the user registers? You can use nodemailer to send emails out and create a verification route in your backend to check and verify tokens

**Comment by talmax1124** [comment]
No, I am not sending emails at all. How would I do that? I am using node, but not nodemailer or any emailing service for the site currently. @AstroProjection

**Comment by talmax1124** [comment]
How would I do it? @AstroProjection

**Comment by AstroProjection** [comment]
Install nodemailer on your backend, use it to send an email to the user with a link that connects to a particular path on your backend, where you will pass in a jwt token you sent by email.

When they access that endpoint with the token in the url it will be used to verify they received the email and the email Exists, then update that info in the back-end.

https://safwan-du16.medium.com/email-verification-with-node-js-and-nodemailer-3a6363b31060

One example, but you can refine it however you want

**Comment by talmax1124** [comment]
In server.js correct? @AstroProjection

**Comment by talmax1124** [comment]
Hey @AstroProjection I tried doing the tutorial and refining it a bit, and I can't get it to work. Can you check my latest code commit. Thnaks.

**Comment by AstroProjection** [comment]
You haven't pushed anything up

**Comment by talmax1124** [comment]
Oh, I realized I did it on my other repo. Let me share you the link to that
repo. I changed the file userController.
https://github.com/talmax1124/cdshop

**Comment by talmax1124** [comment]
I forgot. I did it on another repo of mine: https://github.com/talmax1124/cdshop @AstroProjection

**Comment by AstroProjection** [comment]
Last changes were a month ago?

**Comment by talmax1124** [comment]
Sorry, this is the link: https://github.com/talmax1124/creativeduoshopping-master @AstroProjection

**Comment by talmax1124** [comment]
This is the actual one, the one I last sent. Sorry. @AstroProjection

**Comment by AstroProjection** [comment]
Can't seem to access the page

**Comment by talmax1124** [comment]
Look at my profile talmax1124 and look for repo: creativeduoshopping-master

**Comment by talmax1124** [comment]
Or try checking the link again @AstroProjection

**Comment by talmax1124** [comment]
I got it working! @AstroProjection . This may sound stupid but I almost had paypal for the place-order screen instead of the order screen. Is there a way you can help me changing it? Thanks and have a great day or evening!

**Comment by talmax1124** [comment]
i've tried it again right now and something is not right. Can you help me in this? @AstroProjection

**Comment by talmax1124** [comment]
Nevermind got it working too! i just need help implementing stripe like this https://github.com/stripe-samples/checkout-one-time-payments or https://checkout.stripe.dev/preview @AstroProjection

**Comment by AstroProjection** [comment]
Awesome! see as you keep solving these issues it all starts to fall into place a little better :)

**Comment by AstroProjection** [comment]
i'll look into the stripe implementation and see if I can help you out with it

**Comment by talmax1124** [comment]
Ok. Thanks Alot @AstroProjection !

**Comment by talmax1124** [comment]
Can you do it in my other repo talmax1124/creativeduoshopping-master or something like that. @AstroProjection Thanks!

**Comment by talmax1124** [comment]
Do you know by when you can have it ready? @AstroProjection

**Comment by talmax1124** [comment]
Did you figure out how to do the stripe thing @AstroProjection

**Comment by AstroProjection** [comment]
Sorry, I've been preoccupied with work, couldn't find the time for it, I'll look at it over the weekend

**Comment by talmax1124** [comment]
It's all good! I have been trying to add the payment before the order screen but I had to revert it because It was sending to sandbox URL. If you ever find a fix, I will appreciate it if you could teach me what I did wrong @AstroProjection

**Comment by talmax1124** [comment]
Hello @AstroProjection I just wanted to see if you had time to accomplish the stripe thing?

**Comment by talmax1124** [comment]
Also I found a bug when deleting the reviews. It deletes the rating but not how many review there was. For Example, the photo is after i deleted a review, it stays like there was a review made. @AstroProjection 

![image](https://user-images.githubusercontent.com/61528430/125223890-85a5ab00-e29a-11eb-8e0b-4fce6bc7f4dd.png)'

Update like 2 minutes after:
I fixed it, just made an equation to remove it by 1 since that is how many 1 comment is. In terms of review.

**Comment by AstroProjection** [comment]
Nice! sorry I've been swamped with work this last month, haven't had time to do my own thing, how did the stripe thing go?

**Comment by talmax1124** [comment]
No Luck! I have been working on other things and work as well. I have done major improvements, just the stripe thing is the only thing I couldn't do. I don't know what to do. That's the reason I asked you. @AstroProjection

**Comment by talmax1124** [comment]
@AstroProjection Could you finish the stripe thing?

**Comment by talmax1124** [comment]
@AstroProjection or if you could help me implement Auth0 into the sign in. Or if you can help me implement payment buttons like Apple Pay or Google Pay. I really don't know how to any of these because I'm implementing some stuff. Thanks for the help!

**Comment by talmax1124** [comment]
@AstroProjection Are you currently busy? Could you help me out in my other repository adding edit review but that only that user can edit it?
